applications/asset_tracker_v2/src/modules/debug_module.c

```diff
@@ -0,0 +1,288 @@
+		size_t len = sizeof(data);
+
+		while (memfault_packetizer_get_chunk(data, &len)) {
+			struct debug_module_event *debug_module_event = new_debug_module_event();
+
+			debug_module_event->type = DEBUG_EVT_MEMFAULT_DATA_READY;
+			debug_module_event->data.memfault.len = len;
+
+			BUILD_ASSERT(sizeof(debug_module_event->data.memfault.buf) >= sizeof(data));
+			memcpy(debug_module_event->data.memfault.buf, data,
+			       sizeof(debug_module_event->data.memfault.buf));
+
+			EVENT_SUBMIT(debug_module_event);
+		}
+
+	} else if (data_available) {
+		memfault_zephyr_port_post_data();
+	}
+}
+
+static void add_gps_metrics(uint8_t satellites, uint32_t search_time,
+			    enum gps_module_event_type event)
+{
+	int err;
+
+	switch (event) {
+	case GPS_EVT_DATA_READY:
+		err = memfault_metrics_heartbeat_set_unsigned(
+						MEMFAULT_METRICS_KEY(GpsTimeToFix),
+						search_time);
+		if (err) {
+			LOG_ERR("Failed updating GpsTimeToFix metric, error: %d", err);
+		}
+		break;
+	case GPS_EVT_TIMEOUT:
+		err = memfault_metrics_heartbeat_set_unsigned(
+						MEMFAULT_METRICS_KEY(GpsTimeoutSearchTime),
+						search_time);
+		if (err) {
+			LOG_ERR("Failed updating GpsTimeoutSearchTime metric, error: %d", err);
+		}
+		break;
+	default:
+		LOG_ERR("Unknown GPS module event type");
+		return;
+	}
+
+	err = memfault_metrics_heartbeat_set_unsigned(MEMFAULT_METRICS_KEY(GpsSatellitesTracked),
+						      satellites);
+	if (err) {
+		LOG_ERR("Failed updating GpsSatellitesTracked metric, error: %d", err);
+	}
+
+	memfault_metrics_heartbeat_debug_trigger();
+}
+
+static void memfault_handle_event(struct debug_msg_data *msg)
+{
+	if (IS_EVENT(msg, app, APP_EVT_START)) {
+		/* Register callback for watchdog events. Used to attach Memfault software
+		 * watchdog.
+		 */
+#if defined(CONFIG_WATCHDOG_APPLICATION)
+		watchdog_register_handler(watchdog_handler);
+#endif
+	}
+
+	/* Send Memfault data at the same time application data is sent to save overhead
+	 * compared to having Memfault SDK trigger regular updates independently. All data
+	 * should preferably be sent within the same LTE RRC connected window.
+	 */
+	if ((IS_EVENT(msg, data, DATA_EVT_DATA_SEND)) ||
+	    (IS_EVENT(msg, cloud, CLOUD_EVT_CONNECTED))) {
+		send_memfault_data();
+		return;
+	}
+
+	if ((IS_EVENT(msg, gps, GPS_EVT_TIMEOUT)) ||
+	    (IS_EVENT(msg, gps, GPS_EVT_DATA_READY))) {
+		add_gps_metrics(msg->module.gps.data.gps.satellites_tracked,
+				msg->module.gps.data.gps.search_time,
+				msg->module.gps.type);
+		return;
+	}
+}
+#endif /* defined(CONFIG_MEMFAULT) */
+
+static void message_handler(struct debug_msg_data *msg)
+{
+	if (IS_EVENT(msg, app, APP_EVT_START)) {
+		int err = module_start(&self);
+
+		if (err) {
+			LOG_ERR("Failed starting module, error: %d", err);
+			SEND_ERROR(debug, DEBUG_EVT_ERROR, err);
+		}
+	}
+
+#if defined(CONFIG_MEMFAULT)
+	memfault_handle_event(msg);
+#endif
+}
+
+EVENT_LISTENER(MODULE, event_handler);
+EVENT_SUBSCRIBE_EARLY(MODULE, app_module_event);
```

Comment:
Why does the debug module have to be an early subscriber to all the events? I was assuming that this be the late subscriber because the data processed and generated by this module is not delay sensitive.

Reply:
In some cases, modules consume events. Meaning that they are not carried around to other modules after the respective module that consumed the event has processed it. Subscribing early is just a precaution to ensure that the debug module most likely receives all events before they are potentially discarded.